### DISCUSSION TITLE
General Menu Existing Functionality Questions

### DISCUSSION THREAD

**Discussion by tonygem2022** [iTwin/itwinjs-core#3898] [Q&A]
Team iTwin.

I hope you can point me in the right direction / right code or advise if the below is not possible or is on an existing road map to these general functionality points.

- Within the Section Tools menu, can we turn off one / some of the "Section elements", for example turn off "Plane" and retain others or do we need to turn the entire menu off and recode those we would like to retain?

-  Similarly, are we able to toggle on/off other Menu buttons, for example the "Scope Assembly" in the bottom toolbar?

- How can we enable the user to rotate the "Range" section tool for example? Currently the tool is aligned to the orthogonal grid.  A temp workaround is the use of the "Shape" tool, however, this solution is not a permanent fix.

- Is it possible to enable the use of arrow keys for moving around a model and panning?

- If we wanted to include a "basic site map" in the corner of the screen and show where a user may be located in relation to an overall site as they move around, how might we approach doing this?

We are learning the system and trying to adapt it to our needs.

Thanks   Tony

**Comment by pmconne** [comment]
Hi Tony,

I think most of your questions will require input from folks like @aruniverse, @NancyMcCallB, and/or @kckst8. You're using the iTwin Viewer component, correct?

The [LookAndMoveTool](https://www.itwinjs.org/reference/core-frontend/tools/lookandmovetool/) allows you to use arrow keys (or WASD) to move around a 3d view, similar to controlling a first-person video game.

I have seen demos of applications that display one primary viewport in which the user navigates and a second viewport displaying a "minimap" of the entire model, with a rectangle indicating the area of the model currently visible in the primary viewport. This works well for 2d views. For a 3d view you could do something similar, displaying a fitted top-down view in a floating viewport as your minimap, with a [Marker](https://www.itwinjs.org/reference/core-frontend/views/marker/) indicating the current position within the site.Here's a [similar discussion](https://github.com/iTwin/itwinjs-core/discussions/2231) containing some suggestions. Here's a [sample](https://www.itwinjs.org/sandboxes/NathanElrod/popout-viewport) that shows one way to pop out a second floating viewport. I think that sample precedes the introduction of beta support for [floating viewports](https://www.itwinjs.org/reference/appui-react/contentview/floatingviewportcontentcontrol/) which may be a more appropriate solution.

**Reply by aruniverse** [reply]
> Within the Section Tools menu, can we turn off one / some of the "Section elements", for example turn off "Plane" and retain others or do we need to turn the entire menu off and recode those we would like to retain?

With the default section tools I don't think this is possible. I think you'd have to use the models/categories trees to toggle things off in conjunction with the section tools.

> Similarly, are we able to toggle on/off other Menu buttons, for example the "Scope Assembly" in the bottom toolbar?

If you're using the iTwin Viewer, what version are you using? 2.x or 3.x? If you're using v3, then you can disable the selection scope in the statusbar by setting [selectionScope](https://github.com/iTwin/viewer/blob/master/packages/modules/viewer-react/src/types.ts#L203) to false for the [ViewerStatusbarItemsProvider](https://github.com/iTwin/viewer/blob/master/packages/modules/viewer-react/src/components/app-ui/providers/ViewerStatusbarItemsProvider.tsx#L29)

**Comment by tonygem2022** [comment]
Thank you @pmconne and @aruniverse. We are building on the V3 journey.

We understood V3 presents more adaptability of "standard" functions. 

Unsure what you refer to re "_use the models/categories trees to toggle things off_"? 

I believe it is understandable that if we are not using a particular function, in this case "Plane", we prefer not to keep it hence we want to remove it. Is there a definite workaround? Is there any possibility of including more flexibility with existing menus?

These was no suggested solution to the "Range" query and enabling the possibility of rotation ?

We will look at the suggestions for the last 2 points and hopefully make these work. We appreciate the suggestions.

Thanks. Tony

**Reply by pmconne** [reply]
> Unsure what you refer to re "use the models/categories trees to toggle things off"?

I think @aruniverse might have misunderstood 'turn off one / some of the "Section elements"' to refer to elements in the view -- you're asking to customize the set of UI elements shown in the submenu to for example remove "Plane". (I don't personally know how to do that or if it's currently supported).

@bbastings for rotation in [ViewClipByRangeTool](https://www.itwinjs.org/reference/core-frontend/tools/viewclipbyrangetool/).

**Comment by tonygem2022** [comment]
Thanks @pmconne .

As an update, i think we can park the request to remove the "Plane" element. I seemed to have got it to work in the CAD environment but it cannot seem to "click" to anything in a Reality mesh model environment to make it work? Any thoughts? Is the mesh model obscuring the clicking plane that is possible in a CAD environment perhaps.....?

Regarding the rotation request, refer below. We need the ability to enable the user to rotate the "Range" tool. Not all models are orthogonally aligned.

![Range](https://user-images.githubusercontent.com/108276006/177891936-50910036-d460-451f-a88b-89148b3558ba.jpg)

https://user-images.githubusercontent.com/108276006/177892420-bd179034-edf3-4ca6-bc81-5cb8e9ad596f.mp4

Also, can you help us understand why when we use the "Shape" tool, the clipping box extents in the vertical direction are so far removed from area we are clipping. Refer video below.

https://user-images.githubusercontent.com/108276006/177893662-fa88036e-e9d8-4b21-a673-b7eee9f9e101.mp4

Thanks. Tony

**Reply by tonygem2022** [reply]
Hi iTwin team.
A gentle prompt please if there is any guidance on how to resolve the recent queries?
Thanks in advance. Tony

**Reply by pmconne** [reply]
I am beginning to lose track of which questions still require answers.
> I seemed to have got it to work in the CAD environment but it cannot seem to "click" to anything in a Reality mesh model environment to make it work? Any thoughts? Is the mesh model obscuring the clicking plane that is possible in a CAD environment perhaps.....?

Please explain the problem in detail. It sounds like you are trying to define a clip volume in a view containing only a reality model, and you can't click on the reality model to define your clip points?

> Rotate clip range box

Typically you'd define a shape instead for this purpose, but you say this is only a "temp workaround" -- what issues are you facing?
The clip-by-range tool can align the box to the auxiliary coordinate system if the ACS context lock is enabled. @bbastings could provide details.

> when we use the "Shape" tool, the clipping box extents in the vertical direction are so far removed from area we are clipping.

It looks like the clip-by-shape extrudes your shape to match the vertical dimension of the iModel's [project extents](https://www.itwinjs.org/reference/core-common/imodels/imodel/projectextents/). Your project extents appear to be quite tall. [This sandbox](https://www.itwinjs.org/sandboxes/RaphaelMelancon/Visualize%20Project%20Extents) demonstrates how to visualize the project extents, if you want to verify.

**Reply by tonygem2022** [reply]
Thanks @pmconne.

To clarify, issues still in need of support to close:

1. Plane issue - i will try and record a video for you to see what i refer to. Issue is though  i can only upload a limited size video so will do my best.

In summary though, the Plane element does not seem to work with reality meshes. Look below.

In a CAD model the Plane tool is able to 'click' to a spot / plane.
![Plane Working](https://user-images.githubusercontent.com/108276006/179736355-9b701426-8e5d-40aa-8ffc-8e1cf38fc8d4.jpg)

However for a reality model, in the same location, the plane tool does not seem to work regardless of where i hover the cursor to identify a point for the section plane.
![Plane Not Working](https://user-images.githubusercontent.com/108276006/179736570-dee1878d-7a8c-4f88-b7c5-ca7037e6fcf4.jpg)

2. Yes the issue is a clip-by-range only follows the grid system of the model. We need the ability to have the clip box parallel or at any angle we require to clip the element of interest. In the video above, you note the skid is not orthogonal to the grid system hence the clip doesnt act the way we need. Is there a way to enable rotation of the clip-by-range to occur, perhaps after is created?

3. Yes, there is a tall column within this CAD model and we presumed this was the reason the extents in the vertical direction were so high. If we are selecting an area only, through clip-by-shape, how do we / or why can't the outcome be only having the extents to that shape being clipped? i.e. in the video, whatever the highest vertical dimension/bounds are within the shape clipped?

Thanks .Tony.

**Reply by pmconne** [reply]
Thanks.
1. The clip-by-plane tool requires a normal vector. For elements it can obtain this by querying the backend for the element's geometry. For reality models it has no access to the geometry (it has been uploaded to GPU and erased from CPU memory to conserve RAM).
2. I asked: what is the problem with using clip-by-shape to define a box of the orientation you desire, rather than attempting to use the clip-by-range tool?
3. You could compute the range using the full vertical size of the project extents, then use an ECSql query to compute the union of the ranges of all elements intersecting that range, then intersect the computed range with the unioned element ranges to limit the Z.

I should point out that all of these general-purpose clipping tools are extensible -- you can customize them for whatever suits your specific purposes.
